### PR TITLE
#4809 dprint tensix regs

### PR DESCRIPTION
**Pull request by ihamer-tt**
Initial commit to dprint Tensix destination registers.

- bit_utils.h/cpp + test: to extract n-bit values from 32bit arrays, and reverse (pack)
- U32_ARRAY and TYPED_U32_ARRAY: to DPRINT a bunch of an array of numbers
- dbg_ halt, unhalt and read_dest_acc_row: to read the tensix dest register values

Currently works for dest registers. Remaining:
- [x] Select tile to show (0..7)

Remaining issues for reading srcA/srcB:
- [x] `TTI_SHIFTXB -> TT_SHIFTXB` in common/inc/ckernel_debug.h

The remaining issue for SrcA/B moved to #6195

### CODE REVIEW

**Comment by ihamer-tt** [comment]
Passing post-commits: https://github.com/tenstorrent-metal/tt-metal/actions/runs/8187167716

**Comment by ihamer-tt** [comment]
@pgkeller @davorchap @mywoodstock: can you, please, review?

File: tt_metal/impl/debug/dprint_server.cpp [hunk 1]
```diff
@@ -194,6 +198,61 @@ static void PrintTileSlice(ostream& stream, uint8_t* ptr, int hart_id) {
     }
 } // PrintTileSlice
 
+// Create a float from a given bit pattern, given the number of bits for the exponent and mantissa.
+// Assumes the following order of bits in the input data:
+//   [sign bit][mantissa bits][exponent bits]
+float make_float(uint8_t exp_bit_count, uint8_t mantissa_bit_count, uint32_t data) {
```
Comment (pgkeller) [inline tt_metal/impl/debug/dprint_server.cpp:204]: static?

**Review by pgkeller — COMMENTED** [review]
A couple things:
 - I can't say I fully follow how this works from the diffs
 - I see DPRINT_MATH(DPRINT << xyz) but never did find where DPRINT_MATH was defined, not sure why this is different from the other DPRINTs (wrapping DPRINT w/ another call or macro)
 - that said, I don't have any signifcant concerns w/ this

File: tt_metal/hw/inc/debug/dprint_tensix.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+void dprint_tensix_dest_reg(int tile_id = 0) {
+    dbg_halt();
+    MATH({
+        DPRINT_MATH(DPRINT << FIXED() << SETPRECISION(2));
+        uint32_t rd_data[8+1]; // data + array_type
+        for (int row = 0; row < 64; row++) {
+            dbg_read_dest_acc_row(row + 64 * tile_id, rd_data);
+            DPRINT_MATH(DPRINT << SETW(6) << TYPED_U32_ARRAY(TypedU32_ARRAY_Format_TensixRegister_FP16_B, rd_data, 8));
+            if (row % 2 == 1) DPRINT_MATH(DPRINT << ENDL());
+        }
```
Comment (mywoodstock) [inline tt_metal/hw/inc/debug/dprint_tensix.h:20]: Doesn't `MATH(( ))` already limit compilation to triscs? Then we dont need `DPRINT_MATH`?

**Review by mywoodstock — APPROVED** [review]

File: tt_metal/hw/inc/debug/dprint_tensix.h (hunk 2 above)
Comment (ihamer-tt) [reply tt_metal/hw/inc/debug/dprint_tensix.h:20]: Yes, thanks. Removed.

File: tt_metal/impl/debug/dprint_server.cpp (hunk 1 above)
Comment (ihamer-tt) [reply tt_metal/impl/debug/dprint_server.cpp]: Changed

**Review by acejkov — APPROVED** [review]

**Pull request merged**